feature(mer-49): ILF and HSMs blog (payment HSMs) (#225)
interledger/interledger.org-developers: src/content/blog/2026-05-12-ilf-cards-hsm-integration.md

---
title: 'Signed, Sealed, Hardware Enforced: HSMs for Payment Trust'
description: 'How HSMs protect keys, enforce trust boundaries, and secure card payment flows with Rafiki and Interledger.'
date: 2026-05-12
slug: ilf-and-hsms
authors:
  - Jason Bruwer
author_urls:
  - https://github.com/koekiebox
  - https://www.linkedin.com/in/jason-bruwer-8110766/
tags:
  - Interledger Protocol
  - Card Payments
  - Rafiki
  - Updates
  - HSM
---

> HSMs sit at the heart of modern payment security. Trusted, hardened, and responsible for protecting the cryptographic keys that financial systems rely on. Our next exploration asks a pivotal question: how do HSMs fit into the world of payments, and how can they complement Rafiki and Interledger-based architectures without weakening the trust model that regulated financial infrastructure demands?

Reviewer: ```suggestion
HSMs sit at the heart of modern payment security. Trusted, hardened, and responsible for protecting the cryptographic keys that financial systems rely on. Our next exploration asks a pivotal question: how do HSMs fit into the world of payments, and how can they complement Rafiki and Interledger-based architectures without weakening the trust model that regulates financial infrastructure demands?
```

An HSM is a specialized cryptographic device, or in some cases a tightly controlled managed service, designed to generate, store, protect, and use cryptographic keys without exposing them in clear form to general-purpose application environments.

At a high level, an HSM acts as a hardened trust anchor. Rather than allowing sensitive keys to live in application memory, configuration files, or developer-managed infrastructure, the HSM keeps those keys within a controlled boundary and performs sensitive operations on behalf of other systems.

In practical terms, this means an application might ask an HSM to do things like:

- Generate a symmetric or asymmetric key
- Encrypt or decrypt data
- Sign or verify a message
- Wrap one key under another
- Derive transaction keys
- Rotate or retire cryptographic material

The important detail is that the application may use the key, but it should not need to "see" the key in the clear. **That distinction matters enormously**.

If a normal server is compromised, secrets stored in memory or on disk are often at risk. An HSM is specifically designed to reduce that risk by creating a separate, hardened environment for cryptographic operations.

> In regulated environments, it also helps enforce policies around who can perform which operations, how keys are imported or exported, and what kinds of usage are allowed.

Reviewer: ```suggestion
In regulated environments, it also helps enforce policies around who can perform which operations, how keys are imported or exported, and the usages allowed for individual keys.
```

So while it is tempting to think of an HSM as "just a box that stores keys", that is too narrow. An HSM is better understood as a controlled boundary for trust.

## Why Do We Need an HSM?

> If all we needed were encryption, software libraries would often be enough.

Reviewer: ```suggestion
If all we needed was encryption, software libraries would often be enough.
```

**THAT!** is why HSMs exist.

### Protecting the most sensitive secrets

Master keys, signing keys, derivation keys, and CA keys are "crown jewel material." If they are exposed, entire trust chains can be broken — not just a single request or environment. An HSM reduces that exposure by ensuring such keys are generated and used within a controlled boundary.

### Separating duties and trust boundaries

In real systems, not every service should have equal access to secrets. A payment API may need to request an operation, but it should not be free to extract every key.
An operations team may need to deploy services, but they should not automatically gain access to master key material.
Security teams may need oversight without having to review every transaction manually.

HSMs help enforce these boundaries by moving sensitive operations into a dedicated trust domain.

### Supporting auditability and compliance

In financial environments, "secure enough" is rarely a vague engineering judgment. There are standards, audits, certifications, and contractual expectations.

> Institutions need to show not just that encryption exists, but that key handling follows controlled processes. HSMs support this by providing stronger operational controls, usage policies, dual-control workflows in some deployments, and audit trails around key management.

Reviewer: ```suggestion
Institutions need to show not only that encryption is applied, but that the encryption implementation itself is verified and that the key management follows controlled processes. HSMs support this by providing verified implementations, stronger operational controls, strict usage policies, dual-control workflows in some deployments, and audit trails around key management.
```

### Reducing blast radius

> Even strong applications can have bugs. Even well-managed servers can be compromised. One of the key advantages of an HSM is that it reduces the blast radius when other parts of the environment go wrong. A service might be able to submit a signing request, but not exfiltrate the long-term signing key. A workflow might be able to request a wrapped transaction key, but not obtain the master key used to derive it.

Reviewer: ```suggestion
Even secure applications can have bugs. Even well-managed servers can be compromised. One of the key advantages of an HSM is that it reduces the blast radius when other parts of the environment go wrong. A service might be able to submit a signing request, but not exfiltrate the long-term signing key. A workflow might be able to request a wrapped transaction key, but not obtain the master key used to derive it.
```

That difference can be the line between an incident that is contained and one that becomes systemic.

## Why are HSMs Critical in Payments?

HSMs are relevant in many industries, but payments are one of the clearest examples of where they become indispensable.

This is because payment systems are full of high-value cryptographic operations. They do not just protect data at rest or in transit.
They establish trust between issuers, acquirers, terminals, processors, payment applications, cards, and backend services.

### Payments are built on key hierarchies

Card payments rely on structured key hierarchies and tightly defined cryptographic processes.
There are issuer-side keys, terminal-side keys, transport keys, PIN-related keys, transaction keys, derivation keys, and keys used for encryption, MACing, or signing.

> These are not casual secrets. They determine whether one party can trust another's output.

Reviewer: ```suggestion
These are not casual secrets. They determine whether one party can trust and verify the integrity and authenticity of data received from another party.
```

Please read the [Rafiki Card Integration BLOG](https://interledger.org/developers/blog/rafiki-card-integration/) for background on Rafiki Card Payments if you haven't already.

![ILF + Cards](/developers/img/blog/2026-05-01/hsm.png)

[Rafiki](https://rafiki.dev/) is an open-source platform that enables Account Servicing Entities (ASEs) like banks and digital wallet providers to integrate [Interledger Protocol](/developers/get-started) (ILP) functionality into their systems.

## Table of Contents

1. [Why Hardware Security Modules Matter in Payments and How They Relate to Rafiki](#why-hardware-security-modules-matter-in-payments-and-how-they-relate-to-rafiki)
2. [What Is an HSM?](#what-is-an-hsm)
3. [Why Do We Need an HSM?](#why-do-we-need-an-hsm)
4. [Why Are HSMs Critical in Payments?](#why-are-hsms-critical-in-payments)
5. [How Could the ILF Make Use of HSMs?](#how-could-the-ilf-make-use-of-hsms)
6. [Conclusion](#conclusion)
7. [References](#references)
8. [Glossary of Terms](#glossary-of-terms)

## Why Hardware Security Modules Matter in Payments and How They Relate to Rafiki

> Card payments, digital-wallets, and modern financial APIs all depend on one thing that users rarely see: **trust**.

Reviewer 2: The first mention of "digital wallet" appears without a hyphen, so we should stick to that convention throughout.

Author: Reworked it a bit, fixed.

Not just trust in the institution, or the network, or the device - but trust in the cryptography that protects identities, keys, approvals, and movement of value.

> That **trust** does not happen by accident. It is established through carefully managed cryptographic boundaries, clear ownership of keys, and systems that are designed to avoid exposing secrets where they do not belong. In payment environments especially, this becomes a foundational concern.

Reviewer 2: clear key ownership, and systems designed to avoid exposing secrets

Author: Reworked it a bit, fixed. Please re-review.

This is where Hardware Security Modules, or HSMs, come in.

In our earlier exploration of card payments and Rafiki, a recurring theme emerged: trust is defined as much by key management as by APIs. We looked at POS onboarding, remote key injection, device identity, and separation between payment cryptography and ILP-facing services.

> HSMs sit naturally inside that discussion because they are one of the primary ways financial systems generate, protect, and use sensitive cryptographic material securely.

Reviewer 2: HSMs sit naturally within that discussion because they are among the primary ways

Author: Fixed. Please re-review. 🙇

This post explores what HSMs are, why they matter, why they are so important in payments, and how they can be relevant in architectures that use Rafiki and the Interledger Foundation's broader ecosystem.

## What Is an HSM?

> An HSM is a specialized cryptographic device, or in some cases a tightly controlled managed service, designed to generate, store, protect, and use cryptographic keys without exposing those keys in clear form to general-purpose application environments.

Reviewer 2: wihtout exposing them in clear form

Author: `* wihtout / without` ;-) 
Fixed. Please re-review. 🙇

In practical terms, this means an application might ask an HSM to do things like:

- Generate a symmetric or asymmetric key
- Encrypt or decrypt data
- Sign or verify a message
- Wrap one key under another
- Derive transaction keys
- Rotate or retire cryptographic material

The important detail is that the application may use the key, but it should not need to "see" the key in the clear. **That distinction matters enormously**.

If a normal server is compromised, secrets stored in memory or on disk are often at risk. An HSM is specifically designed to reduce that risk by creating a separate, hardened environment for cryptographic operations.
In regulated environments, it also helps enforce policies around who can perform which operations, how keys are imported or exported, and what kinds of usage are allowed.

So while it is tempting to think of an HSM as "just a box that stores keys", that is too narrow. An HSM is better understood as a controlled boundary for trust.

## Why Do We Need an HSM?

> If all we needed was encryption, software libraries would often be enough.

Reviewer 2: If all we needed were encryption

Author: Fixed. Please re-review. 🙇

Modern cryptographic libraries are powerful, well-tested, and widely available.
But in financial systems, the question is not only whether encryption is possible.
The question is whether sensitive keys can be protected, governed, audited, and used in a way that satisfies both operational reality and security expectations.

**THAT!** is why HSMs exist.

### Protecting the most sensitive secrets

Some keys are simply too important to leave lying around in ordinary infrastructure. Master keys, signing keys, derivation keys, CA keys, and keys used to protect customer or transaction data are often considered "crown jewel material".
If they are exposed, the damage is not limited to one request or one environment. Entire trust chains can be broken.

An HSM reduces that exposure by ensuring such keys are generated and used within a much more controlled environment.

### Separating duties and trust boundaries

In real systems, not every service should have equal access to secrets. A payment API may need to request an operation, but it should not be free to extract every key.

> An operations team may need to deploy services, but they should not automatically gain access to master key material. Security teams may need oversight without manually touching every transaction.

Reviewer 2: Security teams may need oversight without having to review every transaction manually.

Author: Fixed. Please re-review. 🙇

That difference can be the line between an incident that is contained and one that becomes systemic.

## Why are HSMs Critical in Payments?

HSMs are relevant in many industries, but payments are one of the clearest examples of where they become indispensable.

This is because payment systems are full of high-value cryptographic operations. They do not just protect data at rest or in transit.
They establish trust between issuers, acquirers, terminals, processors, payment applications, cards, and backend services.

### Payments are built on key hierarchies

Card payments rely on structured key hierarchies and tightly defined cryptographic processes.
There are issuer-side keys, terminal-side keys, transport keys, PIN-related keys, transaction keys, derivation keys, and keys used for encryption, MACing, or signing.

> These are not casual secrets. They define whether one party can trust the output of another.

Reviewer 2: They determine whether one party can trust another's output.

Author: Fixed. Please re-review. 🙇
